modifiers/ops_abstractor.cpp

```diff
@@ -281,9 +281,23 @@ WalkerStepResult OpsAbstractor::ConcretizationWalker::visit_term(Term & term)
       }
 
       Term rebuilt = solver_->make_term(op, cached_children);
-      // Note: reversed order because update_term_cache takes the
-      // concrete term first
-      oa_.update_term_cache(rebuilt, term);
+
+      if (oa_.abstraction_cache_.find(rebuilt)
+          != oa_.abstraction_cache_.end()) {
+        // Corner case: the abstract term itself is an operation to abstract.
+        // This can occur in invariants derived from the prover.
+        // If the concrete term is already in the cache, calling
+        // `oa_.update_term_cache` triggers an assertion failure.
+        // In this case, only record the abstract-to-concrete term mapping.
+        assert(oa_.ops_to_abstract_.find(op) != oa_.ops_to_abstract_.end()
+               && (rebuilt->get_sort()->get_sort_kind() != BV
+                   || rebuilt->get_sort()->get_width() > oa_.min_bw_));
+        oa_.concretization_cache_[term] = rebuilt;
```

Comment:
https://github.com/stanford-centaur/pono/commit/d8a6eb0303c9c1f1dcb427686a67d01d85110d31#commitcomment-164274465 from @ahmed-irfan:

> how about we update the code of update_term_cache? If a term is present in the concretization/abstraction cache, then the mapped value in the cache should be equal to the one we are currently asking to update. What do you think?

Reply:
@ahmed-irfan I refactored it in 4640748. What do you think?